### PR TITLE
Fix: improve PayPal Donation token handling and refreshing

### PR DESCRIPTION
**Pull request by ravinderk**
## Description

I implement logic to validate the access token when requesting a new access token to PayPal comments. A log entry will generate on failure. The token refresh action will run every 5 minutes until the website get a valid PayPal Commerce access token.

Paypal HTTP client from Paypal SDK forces fresh access tokens on each donation request. I implemented logic to use stored access tokens.

## Affects

This pull request will affect the following:
- PayPal commerce onboarding
- Refresh token access cron job.
- Ontime donation and subscription processing.

## Testing Instructions

- User should be able to onboard.
- Access token should refresh before half an hour of expiration time.
- Donors should be able to create subscriptions.
- Donors should able to donate one-time donations with credit on the PayPal Express account connected website. 

## Additional Changes
I discovered that a seller access token must connect the user to the same PayPal Rest APP. For this reason, we will keep it for future use. The seller access token does not update, but in the future, we can use a refresh token to update the seller access token.

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204019794141811

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@ravinderk It doesn't look like we're "upgrading" the workflows as much as adding additional checks and logging, right? But if something fails it won't somehow recover itself more effectively, right?

**Comment by ravinderk** [comment]
> @ravinderk It doesn't look like we're "upgrading" the workflows as much as adding additional checks and logging, right? But if something fails it won't somehow recover itself more effectively, right?
> 
> I'm just reading the PR title and making sure I'm not missing something when I look at the code.

@JasonTheAdams Yes. This pull request implements better handling of PayPal refresh token response. I am waiting for PayPal guidance to reconnect the existing account.

**Comment by JasonTheAdams** [comment]
@ravinderk Where is this at? Is it waiting for my review?

**Comment by ravinderk** [comment]
> @ravinderk Where is this at? Is it waiting for my review?

@JasonTheAdams Yes. You can review this pull request.

File: src/PaymentGateways/PayPalCommerce/PayPalCheckoutSdk/PayPalHttpClient.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     *
+     * @unreleased x.x.x
+     */
+    private function getAuthorizationInjector($environment, $refreshToken): AuthorizationInjector
+    {
+        $merchant = give(MerchantDetail::class);
+        $authorizationInjector = new AuthorizationInjector($this, $environment, $refreshToken);
+
+        // Set access token if exists.
+        if ($merchant->accessToken) {
```
Comment (ravinderk) [inline src/PaymentGateways/PayPalCommerce/PayPalCheckoutSdk/PayPalHttpClient.php:47]: @JasonTheAdams To force refresh access token on every request for testing, you can either comment this `if` condition or set `false`.

**Review by JasonTheAdams — COMMENTED** [review]
Great work, @ravinderk. Admittedly, it's a bit difficult to follow along with all of this, since we're going in an out of different systems. My biggest concern is that I don't see how the retry flow works, here. As we've discussed, ideally a request would be attempted, then checked to see if it failed specifically due to token failure, and then retried. If that's happening, then we're good to send this to QA for **thorough** testing.

**Comment by ravinderk** [comment]
Note

Asana Task:https://app.asana.com/0/1203193833018151/1204019794141811/f

**Review by Benunc — APPROVED** [review]
This passes automated tests and al manual tests.

**Review by JasonTheAdams — APPROVED** [review]

**Pull request merged**